tmva/pymva/src/MethodPyKeras.cxx

```diff
@@ -154,51 +155,132 @@ UInt_t TMVA::MethodPyKeras::GetNumValidationSamples()
-      PyRunString("import tensorflow as tf","Error importing tensorflow");
-      PyRunString("from keras.backend import tensorflow_backend as K");
-      // run these above lines also in global namespace to make them visible overall
-      PyRun_String("import tensorflow as tf", Py_single_input, fGlobalNS, fGlobalNS);
-      PyRun_String("from keras.backend import tensorflow_backend as K", Py_single_input, fGlobalNS, fGlobalNS);
+void MethodPyKeras::SetupKerasModel(bool loadTrainedModel) {
+
+   // initialize first Keras. This is done only here when class has
+   // all state variable set from options or read from XML file
+   // Import Keras
+
+   bool useTFBackend = kFALSE;
+   bool kerasIsCompatible = kTRUE;
+   bool kerasIsPresent = kFALSE;
+
+   if (!fUseTFKeras) {
+      auto ret  = PyRun_String("import keras", Py_single_input, fGlobalNS, fLocalNS);
+      // need importing also in global namespace
+      if (ret != nullptr) ret = PyRun_String("import keras", Py_single_input, fGlobalNS, fGlobalNS);
+      if (ret != nullptr)
+         kerasIsPresent = kTRUE;
+      if (kerasIsPresent) {
+         // check compatibility with tensorflow
+         if (GetKerasBackend() == kTensorFlow ) {
+            useTFBackend = kTRUE;
+
+            PyRunString("keras_major_version = int(keras.__version__.split('.')[0])");
+            PyRunString("keras_minor_version = int(keras.__version__.split('.')[1])");
+            PyObject *pyKerasMajorVersion = PyDict_GetItemString(fLocalNS, "keras_major_version");
+            PyObject *pyKerasMinorVersion = PyDict_GetItemString(fLocalNS, "keras_minor_version");
+            int kerasMajorVersion = PyLong_AsLong(pyKerasMajorVersion);
+            int kerasMinorVersion = PyLong_AsLong(pyKerasMinorVersion);
+            Log() << kINFO << "Using Keras version " << kerasMajorVersion << "." << kerasMinorVersion << Endl;
+            kerasIsCompatible = (kerasMajorVersion >= 2 && kerasMinorVersion >= 3);
 
+         }
+      } else {
+         // Keras is not found. try tyo use tf.keras
+         Log() << kINFO << "Keras is not found. Trying using tf.keras" << Endl;
+         fUseTFKeras = 1;
+      }
+   }
+
+   // import Tensoprflow (if requested or because is keras backend)
+   if (fUseTFKeras || useTFBackend) {
+      auto ret = PyRun_String("import tensorflow as tf", Py_single_input, fGlobalNS, fLocalNS);
+      if (ret != nullptr) ret = PyRun_String("import tensorflow as tf", Py_single_input, fGlobalNS, fGlobalNS);
+      if (ret == nullptr) {
+         Log() << kFATAL << "Importing tensorflow failed" << Endl;
+      }
       // check tensorflow version
       PyRunString("tf_major_version = int(tf.__version__.split('.')[0])");
       PyObject *pyTfVersion = PyDict_GetItemString(fLocalNS, "tf_major_version");
       int tfVersion = PyLong_AsLong(pyTfVersion);
       Log() << kINFO << "Using Tensorflow version " << tfVersion << Endl;
 
+      if (tfVersion < 2) {
+         if (fUseTFKeras == 1) {
```

Comment:
I think the logic here is not fully correct. What happens if we have `tfVersion < 2` and `fUseTFKeras == 1`? Are we sure that `kerasIsPresent` or `kerasIsCompatible` is false so we error out gracefully?

Reply:
In that case if Keras is present normal Keras with Tensorflow 1 backend is used. If Keras is not present we flag a fatal error

Reply:
I realized the warning/error message is misleading. I will fix it